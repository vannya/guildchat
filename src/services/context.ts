import { createContext } from 'react';

export const AuthContext = createContext({ isAuth: false, userId: null });
