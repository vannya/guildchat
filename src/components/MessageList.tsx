import React, { useRef, useEffect, useContext } from 'react';
import { MessageRow } from './MessageRow';
import { MessageItem } from '../types';
import { AuthContext } from '../services/context';

export interface IMessageListProps {
	messages: MessageItem[];
}

export const MessageList = ({ messages }: IMessageListProps): JSX.Element => {
	const messagesEndRef = useRef(null);
	const scrollToBottom = (): void => {
		messagesEndRef.current?.scrollIntoView({ behavior: 'smooth' });
	};
	const context = useContext(AuthContext);
	useEffect(() => {
		scrollToBottom();
	}, [messages.length]);

	return (
		<div className='message-list'>
			{messages.map(item => {
				return (
					<MessageRow
						key={item.timeStamp.toString()}
						text={item.content}
						displayType={item.senderId === context.userId ? 'outgoing' : 'incoming'}
					/>
				);
			})}
			<div ref={messagesEndRef} />
		</div>
	);
};
