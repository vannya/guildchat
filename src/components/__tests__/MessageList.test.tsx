import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { AuthContext } from '../../services/context';
import { MessageList, IMessageListProps } from '../MessageList';

// Mock scrollIntoView
const scrollIntoViewMock = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

describe('MessageList', () => {
	const getProps = (): IMessageListProps => {
		return {
			messages: [
				{
					senderId: '1',
					timeStamp: new Date('07/28/2021'),
					content: 'Message 1'
				},
				{
					senderId: '2',
					timeStamp: new Date('07/29/2021'),
					content: 'Message 2'
				}
			]
		};
	};

	it('should match snapshot', () => {
		const props = getProps();
		const { container } = render(
			<AuthContext.Provider value={{ isAuth: true, userId: '1' }}>
				<MessageList {...props} />
			</AuthContext.Provider>
		);
		expect(container).toMatchSnapshot();
		cleanup();
	});

	it('should display the messages correctly', () => {
		const props = getProps();
		const { getAllByText } = render(
			<AuthContext.Provider value={{ isAuth: true, userId: '1' }}>
				<MessageList {...props} />
			</AuthContext.Provider>
		);
		const message1 = getAllByText(props.messages[0].content);
		expect(message1).toHaveLength(1);
		expect(message1[0].classList).toContain('message-row--outgoing');
		const message2 = getAllByText(props.messages[1].content);
		expect(message2).toHaveLength(1);
		expect(message2[0].classList).toContain('message-row--incoming');
		cleanup();
	});
});
