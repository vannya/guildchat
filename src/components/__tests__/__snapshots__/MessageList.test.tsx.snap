// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should match snapshot 1`] = `
<div>
  <div
    class="message-list"
  >
    <div
      class="message-row--outgoing"
    >
      Message 1
    </div>
    <div
      class="message-row--incoming"
    >
      Message 2
    </div>
    <div />
  </div>
</div>
`;
